### ISSUE TITLE
Training and using a model for Overlap Detection

### ISSUE THREAD

**Issue by vsakkas96** [pyannote/pyannote-audio#185]
Hi, thanks for making this open source library.

I've been  trying to use `pyannote` in order to detect speech overlaps on my own dataset (which includes some WAV files). So what I'm trying to do is setup `pyannote` and use it to train, validate an overlap detection model and then apply it on my dataset.

I've downloaded `pyannote` and its dependencies, including the AMI and MUSAN datasets. I've also managed to get [this](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/models/pretrained) tutorial up and running with no issues.

I'm now trying to use [this](https://github.com/pyannote/pyannote-audio/blob/master/pyannote/audio/applications/overlap_detection.py) in order to train and evaluate my overlap detection model. I've created a directory named `overlap_experiment` which contains this `config.yml` file:

```yaml
# train the network for overlapping speech detection
    # see pyannote.audio.labeling.tasks for more details
    task:
       name: OverlapDetection
       params:
          duration: 3.2     # sub-sequence duration
          per_epoch: 1      # 1 day of audio per epoch
          batch_size: 32    # number of sub-sequences per batch

    # use precomputed features (see feature extraction tutorial)
    feature_extraction:
       name: Precomputed
       params:
          root_dir: tutorials/feature-extraction

    # use the StackedRNN architecture.
    # see pyannote.audio.labeling.models for more details
    architecture:
       name: StackedRNN
       params:
         rnn: LSTM
         recurrent: [32, 20]
         bidirectional: True
         linear: [40, 10]

    # use cyclic learning rate scheduler
    scheduler:
       name: CyclicScheduler
       params:
           learning_rate: auto
```

The `overlap_experiment` directory is in the same directory with the `pyannote` and `tutorials` directories. 

My `dataset.yml` file is the following:

```yaml
Databases:
  AMI: /dev/pyannote-audio/amicorpus/*/audio/{uri}.wav
  MUSAN: /dev/pyannote-audio/musan/*/*/{uri}.wav
```

I've used the following command in order to see how to train and use my overlap detection model:
```bash
pyannote-overlap-detection --help
```

Which works, so the overlap detection command can successfully be found.

I then try to run this command:
```bash
pyannote-overlap-detection train ~/dev/pyannote-audio/overlap_experiment AMI.SpeakerDiarization.MixHeadset
```
Where `~/dev/pyannote-audio/overlap_experiment` is the absolute path to my experiment directory which also contains the `config.yml` file I mentioned earlier. However, once I run this command, I get the following error:

```bash
Traceback (most recent call last):
  File "/home/bill/dev/pyannote-audio/venv/bin/pyannote-overlap-detection", line 10, in <module>
    sys.exit(main())
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/applications/overlap_detection.py", line 313, in main
    training=True)
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/applications/base_labeling.py", line 46, in __init__
    experiment_dir, db_yml=db_yml, training=training)
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/applications/base.py", line 178, in __init__
    augmentation=augmentation,
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/features/precomputed.py", line 124, in __init__
    dimension = len(labels)
TypeError: object of type 'NoneType' has no len()
```
So far I haven't managed to find why. In order to see if there is anything wrong with my current setup, I also tried following [this](https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/pipelines/speech_activity_detection) tutorial. I ran the following commands which worked fine:

```bash
cat tutorials/pipelines/speech_activity_detection/config.yml
export EXPERIMENT_DIR=tutorials/pipelines/speech_activity_detection
```

But when I run this:

```bash
pyannote-pipeline train --forever ${EXPERIMENT_DIR} AMI.SpeakerDiarization.MixHeadset
```
I get the same error:

```bash
Traceback (most recent call last):
  File "/home/bill/dev/pyannote-audio/venv/bin/pyannote-pipeline", line 10, in <module>
    sys.exit(main())
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/pipeline/experiment.py", line 390, in main
    experiment = Experiment(experiment_dir, training=True)
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/pipeline/experiment.py", line 180, in __init__
    self.pipeline_ = Klass(**self.config_['pipeline'].get('params', {}))
  File "/home/bill/dev/pyannote-audio/pyannote/audio/pipeline/speech_activity_detection.py", line 81, in __init__
    self._precomputed = Precomputed(self.scores)
  File "/home/bill/dev/pyannote-audio/pyannote/audio/features/precomputed.py", line 124, in __init__
    dimension = len(labels)
TypeError: object of type 'NoneType' has no len()
```

The next steps of the tutorial, such as using the `apply` flag also produce the same error.

Could you perhaps guide me on how to get the overlap detection running? I think it would be great to go step by step on how to use this library for overlap detection and it would help future users as well.

Thanks.

**Update**

I found out that the `root_dir` flag of the `config.yml` file that I created for the experiment should be:

```yaml
root_dir: tutorials/feature_extraction
```

and NOT:

```yaml
root_dir: tutorials/feature-extraction
```
This resulted in the `procomputed.py` file not being able to find the directory. However, that still doesn't fix the issue I mentioned above, as it seems (from *line 91*) that the `metadata.yml` file does not exist in that directory.

**Update 2**

I added this `metadata.yml` file that I found in order to see what happens next:

```yaml
dimension: 2
duration: 0.025
labels:
- non_speech
- speech
start: -0.0125
step: 0.01
```

The above errors stops appearing but I now get the following:

```bash
Traceback (most recent call last):
  File "/home/bill/dev/pyannote-audio/venv/bin/pyannote-overlap-detection", line 10, in <module>
    sys.exit(main())
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/applications/overlap_detection.py", line 333, in main
    experiment_dir, db_yml=db_yml, training=True
  File "/home/bill/dev/pyannote-audio/venv/lib/python3.6/site-packages/pyannote/audio/applications/base_labeling.py", line 63, in __init__
    self.get_model_ = partial(Architecture, **params)
TypeError: type object argument after ** must be a mapping, not NoneType
```

It appears that it's because of the `config.yml` file I posted earlier. The `params` field has no value! What should I set it to?

**Comment by hbredin** [comment]
It might just be a problem with the formatting of your `config.yml`. 
Can you try adding one more space before `rnn`, `recurrent`, `bidirectional`, and `linear`?

If that solves your problem, then I think you found a bug (or a feature, who knows?) in `pyYAML` library that is used for parsing `config.yml`. 

Also, I just created a [new branch](https://github.com/pyannote/pyannote-audio/tree/tutorial_overlap/tutorials/models/overlap_detection) with a tutorial for overlap detection.
I would be grateful if you could follow it and let me know of any problem you might encounter -- before I merge it in the main branch.

**Comment by vsakkas96** [comment]
Hi, sorry for the late response and thanks for the answer.

I've managed to make process with training the model for overlap detection. For future reference, I should point out two mistakes that were made on my initial attempts.

Firstly, my `database.yml` should be like this:
```yaml
Databases:
  AMI: /home/myname/dev/pyannote-audio/amicorpus/*/audio/{uri}.wav
  MUSAN: /home/myname/dev/pyannote-audio/musan/{uri}.wav
```
As my initial paths were incorrect. (`dev` is simply a directory where I cloned this repo to)

Secondly, you were right about the missing space characters on the `config.yml` file. Adding one more space fixed yet another issue I was facing.

Lastly, I followed your tutorial in the new branch you made (thanks for making it in such short time) and so far everything seems to be working fine!
